### PR TITLE
HPCC-30599 Fix file access costs for keyed join

### PR DESCRIPTION
**Pull request by shamser**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

File: system/jlib/jstats.cpp
Comment (shamser) [inline system/jlib/jstats.cpp:1325]: 2 stat kinds found missing in jhtree: StTimeNodeLoad and StTimeNodeRead

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1291]: I think the conclusion from our offline discussion, was that this contextLogger capture/updating of inactiveStats should become redundant.
Because, you are now each handler has its own CStatsContextLogger, which the IKeyManager's are using, and recording jhtree stats to, and you are harvesting into 'fileStats'.

This process is still recording them into the activity stats (into inactiveStats), but I think we said it should not need to anymore, and instead they should be totaled (in the worker), from the collected fileStats.

However, it probably is worth keeping like this for this PR now, and removing the duplicated processes/code in a subsequent PR.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1389]: There's another problem here.
A lookup handler may be dealing with >1 part, but there is only one CStatsContextLogger involved here.
Which means in getFileStats below, it will iterate through the IKeyManager's (1 per part), and merge stats from the shared context logger used by all, causing stats to be over reported.

**Review by jakesmith — COMMENTED** [review]
@shamser - please see comments.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1372]: trivial: perhaps this should call reset() in each (woudl need a CStatsContextLogger::reset added).

**Review by jakesmith — COMMENTED** [review]
@shamser - looks good in general.

I few minor issues only within KJ I think.

However, I think there's at least 1 change, that should be extracted out as an initial PR before this PR (see comment : https://github.com/hpcc-systems/HPCC-Platform/pull/18023#discussion_r1403184462).
That will help reduce the scope/size of the actual KJ change.

Also, it looks like there may be some double counting issues in index read, since there's only one contextLogger shared amongst parts (you may not notice unless the index read was reading >1 part, but it also looked like it was double counting IO stats.
I've opened a new JIRA re. creating some regression tests that grab stats and output them as results, to get some coverage of stat. changes: 

Could you also update the PR and commit message, to say what changes have been made and why, e.g. the removal of the deltaUpdater, the introduction of a CStatsContextLogger per part, contained in the handlers etc, and the reasoning.
We may know the reasoning, but it would help others if description explained.

File: system/jlib/jstats.cpp
Comment (shamser) [reply system/jlib/jstats.cpp:1325]: This line is actually shorter than the previous 3 lines.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1372]: The contextLoggers will have an empty array when it is first called.  This loop pushing back parts.size() items to the array is needed somewhere.  After adding the elements to the array, it's not worth calling reset for each item....should it be setting up the array elsewhere and calling reset on each item here?

**Comment by shamser** [comment]
@jakesmith I've fixed the remote key lookup stats.  jobstats shows consistent jhtreecache stats between local and remote.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1373]: recent change: 

- jhtree and io stats are returned from CKeyLookupLocalHandler::getFileStats().  It uses keyManager->mergeStats()
- some jhtree and io stats are returned CRemoteLookupHandler::getFileStats(). 
- io stats are returned from CFetchLocalLookupHandler::getFileStats

The original point is still valid: contextLoggers only track jhtree stats, so they don't need indexReadFileStatistics.

File: thorlcr/slave/slavmain.cpp
Comment (shamser) [inline thorlcr/slave/slavmain.cpp:762]: Bye the way, this section didn't work before. There was multiple threads were using the same contextLogger.

File: ecl/hthor/hthor.cpp
Comment (shamser) [reply ecl/hthor/hthor.cpp]: That's correct.  The existing code will cause the file cost based on jhtree cache stats to be overwritten by the legacy code.  I've removed the code change to store the read & write cost (to avoid dynamic recalculation) from this PR and put in a separate one: https://track.hpccsystems.com/browse/HPCC-30911

File: thorlcr/activities/indexread/thindexreadslave.cpp
Comment (shamser) [reply thorlcr/activities/indexread/thindexreadslave.cpp]: I've amended this code so it uses multiple contextLoggers.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp]: no. Please see previous explanation https://github.com/hpcc-systems/HPCC-Platform/pull/18023/files#r1406087606

File: thorlcr/activities/indexread/thindexreadslave.cpp
Comment (shamser) [reply thorlcr/activities/indexread/thindexreadslave.cpp]: Jira is not needed.  I have amended this code to collects file stats in gatherActiveStats and, so should be intermittently serialized to master.

File: thorlcr/graph/thgraphmaster.cpp
Comment (shamser) [reply thorlcr/graph/thgraphmaster.cpp]: If we do create a "@legacyReadCost", it wouldn't be exposed to the end-users in ECL watch, limiting it's usefulness. Secondly, the legacy files do not have jhtree stats so it would not be possible to produce an accurate access cost figure.  We could reset the readCost to zero but that may also be misleading.  Should we create new jira to discuss and handle this separately.  (Note: all changes related to handling legacy file costs and dynamic access cost calculation is no longer in this jira.  These changes have been moved to another jira: https://track.hpccsystems.com/browse/HPCC-30911)

Comment (jakesmith) [reply thorlcr/graph/thgraphmaster.cpp]: Will add comments to https://github.com/hpcc-systems/HPCC-Platform/pull/18080

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1373]: Is this related to the offline conversation we had, where the conclusion was that we don't currently collect and publish the io stats, either from the local handler, or from the remote - so we don't want to fix it 1 side now ?
But should fix in a subsequent PR so io stats are collected, and published whether using a local or remote handler.

**Review by jakesmith — COMMENTED** [review]
> @jakesmith this PR is based on https://track.hpccsystems.com/browse/HPCC-30911. The first 2 commits shown in github, is related to https://track.hpccsystems.com/browse/HPCC-30911.


I'll hold off reviewing this further until the PR for 30911 is merged.
Please rebase this PR and tag me, when that is merged.

**Review by jakesmith — COMMENTED** [review]
@shamser - I haven't re-reviewed the keyedjoin changes yet, I had some comments on indexread changes,
and I think it would be best, unless there's a blocking reason not to, to split those changes into a separate PR.
Will discuss offline.

File: thorlcr/activities/indexread/thindexreadslave.cpp
Comment (shamser) [reply thorlcr/activities/indexread/thindexreadslave.cpp]: indexreadslave changes will be moved to separate PR.

File: thorlcr/activities/indexread/thindexreadslave.cpp
Comment (shamser) [reply thorlcr/activities/indexread/thindexreadslave.cpp]: indexreadslave changes will be moved to separate PR.

File: thorlcr/activities/indexread/thindexreadslave.cpp
Comment (shamser) [reply thorlcr/activities/indexread/thindexreadslave.cpp]: indexreadslave changes will be moved to separate PR.

File: thorlcr/activities/indexread/thindexreadslave.cpp
Comment (shamser) [reply thorlcr/activities/indexread/thindexreadslave.cpp]: indexreadslave changes will be moved to separate PR.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:2109]: can you open a JIRA and link it to this JIRA ?

File: thorlcr/graph/thgraphmaster.cpp
Comment (jakesmith) [inline thorlcr/graph/thgraphmaster.cpp:652]: was this comment accidentally reverted? (the one on deleted line 652 seems correct)

File: thorlcr/graph/thgraphmaster.cpp
Comment (jakesmith) [inline thorlcr/graph/thgraphmaster.cpp:692]: trivial: indexread will always be i == 0, only keyedjoin has >1, so slightly more accurate to be:
```
if ((TAKindexread == activityKind) || ((TAKkeyedjoin == activityKind)  && (0 == i)))
```

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1455]: can you open a JIRA and link it to this one?

**Review by jakesmith — COMMENTED** [review]
@shamser - looks good, one question re. blob stats, and some minors + comments re. opening 'future' JIRAs

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1311]: it looks like the remote side (slavmain) would return track and serialize blob stats.
Please test an example to verify same whether using local or remote. Regression test 'indexblobs.ecl' may work as a test query.

Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1311]: Blob stats are not being tracked for keyed join.  However, the blob stats for keyed join was never correct because slavmain was not tracking blob stats.  Also, hthor was not tracking blob stats.  It would be better to address this in a separate jira as the changes affect not only thor  but roxie and hthor too. TODO: create ticket once jira is working again.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:2109]: Will do. TODO: create ticket once jira is working again.

File: thorlcr/graph/thgraphmaster.cpp
Comment (shamser) [reply thorlcr/graph/thgraphmaster.cpp]: Yes, fixed now.

File: thorlcr/graph/thgraphmaster.cpp
Comment (shamser) [reply thorlcr/graph/thgraphmaster.cpp]: Agree, trivial but I've changed the code.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1455]: Will do. TODO: create ticket once jira is working again.

Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1455]: jira: https://track.hpccsystems.com/browse/HPCC-30599

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:2109]: jira: https://track.hpccsystems.com/browse/HPCC-31000

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1311]: jira: https://track.hpccsystems.com/browse/HPCC-31002

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1380]: is this correct? has it been tested? 
i.e. what happens when this is called (because KJ is in a child query or loop), are the total stats correct?

**Review by jakesmith — COMMENTED** [review]
@shamser - looks good, 1 question re. reset() of contextLogger's

**Review by jakesmith — COMMENTED** [review]
@shamser - looks good. Please squash.

Do the remote vs local handler stats. (for index seek/scans) match for the CQ test query ?

**Comment by shamser** [comment]
> @shamser - looks good. Please squash.
> 
> Do the remote vs local handler stats. (for index seek/scans) match for the CQ test query ?

seeks and scans match.

CacheHits is greater (202,062 vs 159499) on non-remote lookup.  NumDiskReads is greater on non-remote (16 vs 12).  Would you expect these to match?

**Comment by jakesmith** [comment]
>CacheHits is greater (202,062 vs 159499) on non-remote lookup. NumDiskReads is greater on non-remote (16 vs 12). Would you expect these to match?

not sure either (both local and remote) are deterministic, because there could be multiple handlers or multiple parts being dealt with by the same process (with shared jhtree cache). However, if N part key, on a N-way Thor - then it may be deterministic..

I'm unsure, I think we can investigate these differences post this PR though.

**Review by jakesmith — COMMENTED** [review]
@shamser - there are some build problems reported in the tests - I think needs rebasing.

**Review by jakesmith — APPROVED** [review]
@shamser - looks good.

File: thorlcr/graph/thgraphmaster.cpp
Comment (ghalliday) [inline thorlcr/graph/thgraphmaster.cpp:669]: Should this code be setting curDiskReads?  It will affect the number of reads recorded against the file in line 675.  (It would also allow the cost calculation to be shared.)

**Review by ghalliday — COMMENTED** [review]
@shamser looks good.  A couple of questions - please could you answer them.  Even if they require changes it is probably best to merge as-is and then revisit as separate PRs.

File: thorlcr/slave/slavmain.cpp
Comment (shamser) [reply thorlcr/slave/slavmain.cpp:776]: Yes, it would be.  I was thinking about doing this separately later as it was not related to kj costs.

File: thorlcr/graph/thgraphmaster.cpp
Comment (shamser) [reply thorlcr/graph/thgraphmaster.cpp:669]: I recall @jakesmith mentioned that it would be useful to have both: the original NumDiskReads and the jhtree stats.

File: system/jhtree/jhtree.cpp
Comment (shamser) [reply system/jhtree/jhtree.cpp:580]: I'll change this.

Comment (shamser) [reply system/jhtree/jhtree.cpp]: Improvement: https://track.hpccsystems.com/browse/HPCC-31035

File: thorlcr/graph/thgraphmaster.cpp
Comment (jakesmith) [reply thorlcr/graph/thgraphmaster.cpp:669]: I think it is fine as it stands.
As it is, we have 2 pieces of info., we have the number of attempts to read from disk (StNumDiskReads) and the number of actual disk reads, at least in the index case (StNumNodeDiskFetches+StNumLeafDiskFetches+StNumBlobDiskFetches).

StNumDiskReads is a tad misleading (and always was), and may not closely correspond to actual disk reads in flat files in some cases either. 
The explanation for what it is (and other stats) is something that should be clarified somewhere, but not directly related to this PR I would say.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**
